# Homework Assignment - Lesson 6: Combining Datasets - Joins

**Due Date:** [Insert Due Date Here]

**Instructions:**
- Complete the following tasks in this R notebook
- Use the exact variable names specified for each task
- Ensure your code runs without errors and produces the expected outputs
- Submit your completed notebook with all code and outputs

---

## Part 1: Data Import and Setup

**Task:** Import the following CSV files and examine their structure.

**Required Variable Names:**
You must use these EXACT variable names:
- `customers` for customers.csv
- `orders` for orders.csv  
- `order_items` for order_items.csv
- `products` for products.csv
- `suppliers` for suppliers.csv

**Expected Outputs:**
- All datasets loaded successfully
- Print dataset dimensions using `nrow()` and `ncol()`
- Show first 3 rows of each dataset using `head(dataset, 3)`

In [1]:
# 1. Load the tidyverse package
library(tidyverse)

# 2. Set working directory to sample_datasets
setwd("../../data")  # Adjust path if needed

# 3. Import all CSV files using read_csv() with EXACT variable names above
customers <- read_csv("customers.csv")
orders <- read_csv("orders.csv")
order_items <- read_csv("order_items.csv")
products <- read_csv("products.csv")
suppliers <- read_csv("suppliers.csv")

# 4. Print dimensions and head() for each dataset
cat("Dataset Dimensions:\n")
cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
cat("Order Items:", nrow(order_items), "rows x", ncol(order_items), "columns\n")
cat("Products:", nrow(products), "rows x", ncol(products), "columns\n")
cat("Suppliers:", nrow(suppliers), "rows x", ncol(suppliers), "columns\n")

# Show first 3 rows of each dataset
head(customers, 3)
head(orders, 3)
head(order_items, 3)
head(products, 3)
head(suppliers, 3)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.2
✔ ggplot2   4.0.0     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.1.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
Rows: 100 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): Name, Email, City
dbl  (1): CustomerID
date (1): Registration_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 250 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl  (3): 

Dataset Dimensions:
Customers: 100 rows x 5 columns
Orders: 250 rows x 4 columns
Order Items: 400 rows x 4 columns
Products: 50 rows x 4 columns
Suppliers: 10 rows x 3 columns


CustomerID Name       Email               City        Registration_Date
1 1          Customer 1 customer1@email.com Phoenix     2020-10-03       
2 2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
3 3          Customer 3 customer3@email.com Chicago     2021-04-20

OrderID CustomerID Order_Date Total_Amount
1 1       87         2023-08-30 424.30      
2 2       12         2024-03-24 183.09      
3 3       37         2024-03-19 549.07

OrderID ProductID Quantity Unit_Price
1 213      8        1         43.81    
2 176     18        5        489.16    
3 118      2        5        442.09

ProductID Product_Name Category    Supplier_ID
1 1         Product 1    Home        6          
2 2         Product 2    Electronics 7          
3 3         Product 3    Home        1

Supplier_ID Supplier_Name Country
1 1           Supplier 1    USA    
2 2           Supplier 2    USA    
3 3           Supplier 3    Japan

## Part 2: Basic Joins

**Tasks:**
1. **Inner Join:** Create `customer_orders_inner` by joining customers and orders
2. **Left Join:** Create `customer_orders_left` to show all customers  
3. **Right Join:** Create `customer_orders_right` to show all orders
4. **Full Join:** Create `customer_orders_full` for complete data view

Use the exact variable names specified above for autograding. Analyze row counts and explain results.

In [2]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
customer_orders <- inner_join(customers, orders, by = "CustomerID")
cat("Inner Join Results:\n")
cat("Customers:", nrow(customers), "\n")
cat("Orders:", nrow(orders), "\n")
cat("Inner Join Result:", nrow(customer_orders), "\n")

# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)
customer_orders_left <- left_join(customers, orders, by = "CustomerID")
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:\n")
cat("Total rows:", nrow(customer_orders_left), "\n")
cat("Customers without orders:", customers_without_orders, "\n")

# Part 2.3: Right Join (REQUIRED variable name: customer_orders_right)
customer_orders_right <- right_join(customers, orders, by = "CustomerID")
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Right Join Results:\n")
cat("Total rows:", nrow(customer_orders_right), "\n")
cat("Orders with invalid customer IDs:", orders_invalid_customers, "\n")

# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
customer_orders_full <- full_join(customers, orders, by = "CustomerID")
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:\n")
cat("Total rows:", nrow(customer_orders_full), "\n")
cat("Customers without orders:", customers_only, "\n")
cat("Orders without valid customers:", orders_only, "\n")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 
Left Join Results:
Total rows: 200 
Customers without orders: 0 


Right Join Results:
Total rows: 250 
Orders with invalid customer IDs: 50 
Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


In [3]:
# Create a left join to keep all customers
customer_orders_left <- left_join(customers, orders, by = "CustomerID")

# Count how many customers have no orders
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))


In [4]:
# Create a right join to keep all orders
customer_orders_right <- right_join(customers, orders, by = "CustomerID")

# Check for orders with invalid customer IDs
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))

In [5]:
## Part 2: Basic Joins

##**Required Variable Names:**

##You must create these EXACT variable names:


# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)

customer_orders <- inner_join(customers, orders, by = "CustomerID")

cat("Inner Join Results:\n")

cat("Customers:", nrow(customers), "\n")

cat("Orders:", nrow(orders), "\n")

cat("Inner Join Result:", nrow(customer_orders), "\n")



# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)

customer_orders_left <- left_join(customers, orders, by = "CustomerID")

customers_without_orders <- sum(is.na(customer_orders_left$OrderID))

cat("Left Join Results:\n")

cat("Total rows:", nrow(customer_orders_left), "\n")

cat("Customers without orders:", customers_without_orders, "\n")



# Part 2.3: Right Join (REQUIRED variable name: customer_orders_right)

customer_orders_right <- right_join(customers, orders, by = "CustomerID")

orders_invalid_customers <- sum(is.na(customer_orders_right$Name))

cat("Right Join Results:\n")

cat("Total rows:", nrow(customer_orders_right), "\n")

cat("Orders with invalid customer IDs:", orders_invalid_customers, "\n")



# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)

customer_orders_full <- full_join(customers, orders, by = "CustomerID")

customers_only <- sum(is.na(customer_orders_full$OrderID))

orders_only <- sum(is.na(customer_orders_full$Name))

cat("Full Join Results:\n")

cat("Total rows:", nrow(customer_orders_full), "\n")

cat("Customers without orders:", customers_only, "\n")

cat("Orders without valid customers:", orders_only, "\n")



##**Expected Outputs:**

##or each join, print:

##- Number of rows in the result using `nrow()`

##- Comparison with original dataset sizes

##- Analysis of unmatched records

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 
Left Join Results:
Total rows: 200 
Customers without orders: 0 
Right Join Results:
Total rows: 250 
Orders with invalid customer IDs: 50 
Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


In [6]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:


# Required output for autograding:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [7]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:


# Required output for autograding:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [8]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:


# Expected output:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [9]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:


# Expected output:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [10]:
# Part 2.3: Right Join (REQUIRED variable name: customer_orders_right)
# Your code here:


# Expected output:
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Right Join Results:
")
cat("Total rows:", nrow(customer_orders_right), "
")
cat("Orders with invalid customer IDs:", orders_invalid_customers, "
")

Right Join Results:
Total rows: 250 
Orders with invalid customer IDs: 50 


In [11]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:


# Expected output:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


In [12]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:


# Required output for autograding:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


## Part 3: Multi-Table Joins

**Tasks:**
1. Create a comprehensive dataset by joining `orders`, `customers`, and `order_items`
2. Extend this dataset by adding `products` and `suppliers` information
3. Create a complete supply chain view showing the full customer-to-supplier relationship

Build your joins step by step and examine the results at each stage.

## Part 3: Multi-Table Joins

**Requirements for Autograding:**
You must create these EXACT variable names in order:

1. `orders_items` - Join orders and order_items  
2. `orders_customers_items` - Add customers to above
3. `complete_order_data` - Add products information
4. `complete_data` - Add suppliers for complete supply chain view

**Expected Outputs:**
Print the number of rows for each step to show the join progression.

In [13]:
# Step 1: REQUIRED variable name: order_items
# Join orders with order_items by OrderID (keep all orders)
order_items <- orders %>%
  left_join(order_items, by = "OrderID")

In [14]:
# Step 1: REQUIRED variable name: order_customer_items
# Join orders and order_items to create the base dataset
order_customer_items <- orders %>%
  left_join(order_items, by = "OrderID")

# Required output for autograding:
cat("Step 1 - Orders with Items:", nrow(order_customer_items), "rows\n")

Step 1 - Orders with Items: 453 rows


In [15]:
# Step 3: REQUIRED variable name: complete_order_data
# Add products information to the order_customer_items dataset
complete_order_data <- order_customer_items %>%
  left_join(products, by = "ProductID")

# Required output for autograding:
cat("Step 3 - Add Products:", nrow(complete_order_data), "rows\n")

Step 3 - Add Products: 453 rows


In [16]:
# Step 4: REQUIRED variable name: complete_data
# Add suppliers information to complete the supply chain view
complete_data <- complete_order_data %>%
  left_join(suppliers, by = "Supplier_ID")

# Required output for autograding:
cat("Step 4 - Complete Supply Chain:", nrow(complete_data), "rows\n")

Step 4 - Complete Supply Chain: 453 rows


## Part 4: Data Quality Analysis

**Tasks:**
1. Use `anti_join()` to find unmatched records between tables
2. Use `semi_join()` to find matched records
3. Check for duplicate keys and analyze their impact on joins
4. Identify and document any data quality issues

This analysis helps understand data integrity and potential issues in your datasets.

In [17]:
# REQUIRED variable name: customers_no_orders
customers_no_orders <- anti_join(customers, orders, by = "CustomerID")
cat("Customers who never placed an order:", nrow(customers_no_orders), "\n")

# REQUIRED variable name: orphaned_orders
orphaned_orders <- anti_join(orders, customers, by = "CustomerID")
cat("Orders without corresponding customers:", nrow(orphaned_orders), "\n")

# REQUIRED variable name: products_never_ordered
products_never_ordered <- anti_join(products, order_items, by = "ProductID")
cat("Products that were never ordered:", nrow(products_never_ordered), "\n")

# REQUIRED variable name: active_customers
active_customers <- semi_join(customers, orders, by = "CustomerID")
cat("Customers who placed at least one order:", nrow(active_customers), "\n")

Customers who never placed an order: 0 
Orders without corresponding customers: 50 
Products that were never ordered: 0 
Customers who placed at least one order: 100 


In [18]:
# REQUIRED variable name: customers_no_orders
customers_no_orders <- anti_join(customers, orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who never placed an order:", nrow(customers_no_orders), "\n")

Customers who never placed an order: 0 


In [19]:
# REQUIRED variable name: orphaned_orders
orphaned_orders <- anti_join(orders, customers, by = "CustomerID")

# Required output for autograding:
cat("Orders without corresponding customers:", nrow(orphaned_orders), "\n")

Orders without corresponding customers: 50 


In [20]:
# REQUIRED variable name: products_never_ordered
products_never_ordered <- anti_join(products, order_items, by = "ProductID")


# Required output for autograding:
cat("Products that were never ordered:", nrow(products_never_ordered), "\n")

Products that were never ordered: 0 


In [21]:
# REQUIRED variable name: active_customers
active_customers <- semi_join(customers, orders, by = "CustomerID")


# Required output for autograding:
cat("Customers who placed at least one order:", nrow(active_customers), "\n")

Customers who placed at least one order: 100 


In [22]:
# Data integrity checks - count duplicate keys

duplicate_customers <- sum(duplicated(customers$CustomerID))
duplicate_orders <- sum(duplicated(orders$OrderID))
duplicate_products <- sum(duplicated(products$ProductID))
duplicate_suppliers <- sum(duplicated(suppliers$Supplier_ID))
# Required outputs for autograding (uncomment after creating the calculations above):
# duplicate_customers <- sum(duplicated(customers$CustomerID))
# duplicate_orders <- sum(duplicated(orders$OrderID))
# duplicate_products <- sum(duplicated(products$ProductID))
# duplicate_suppliers <- sum(duplicated(suppliers$Supplier_ID))

# cat("Data Quality Summary:\n")
# cat("Duplicate customer IDs:", duplicate_customers, "\n")
# cat("Duplicate order IDs:", duplicate_orders, "\n")
# cat("Duplicate product IDs:", duplicate_products, "\n")
# cat("Duplicate supplier IDs:", duplicate_suppliers, "\n")

In [23]:
# REQUIRED variable name: customer_metrics
# Your code here: Calculate customer lifetime value metrics
# Required columns: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value

suppressPackageStartupMessages(library(dplyr))

# Define std_name helper (so the function exists in this cell)
std_name <- function(df, mapping) {
  for (tgt in names(mapping)) {
    cands <- mapping[[tgt]]
    hit <- cands[cands %in% names(df)]
    if (length(hit) > 0 && !(tgt %in% names(df))) {
      names(df)[match(hit[1], names(df))] <- tgt
    }
  }
  df
}

# Normalize column names (no-op if already standard)
orders <- std_name(orders, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  CustomerID = c("CustomerID","Customer_Id","Customer_ID","customer_id","CustomerId")
))
customers <- std_name(customers, list(
  CustomerID = c("CustomerID","Customer_Id","Customer_ID","customer_id","CustomerId"),
  Name       = c("Name","CustomerName","Customer_Name")
))
order_items <- std_name(order_items, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  Quantity   = c("Quantity","Qty","quantity"),
  Unit_Price = c("Unit_Price","UnitPrice","Price","UnitPriceUSD")
))

# Build customer_metrics with required columns
customer_metrics <- customers %>%
  select(CustomerID, Name) %>%
  left_join(orders %>% select(OrderID, CustomerID), by = "CustomerID") %>%
  left_join(order_items %>% select(OrderID, Quantity, Unit_Price), by = "OrderID") %>%
  mutate(line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)) %>%
  group_by(CustomerID, Name) %>%
  summarise(
    Total_Spent = sum(line_total, na.rm = TRUE),
    Order_Count = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  mutate(Avg_Order_Value = ifelse(Order_Count > 0, Total_Spent / Order_Count, 0)) %>%
   arrange(desc(Total_Spent))

# Required output for autograding:
cat("Customer Analysis Summary:
")
cat("Total customers analyzed:", nrow(customer_metrics), "
")
cat("Top customer total spent: $", round(max(customer_metrics$Total_Spent, na.rm = TRUE), 2), "
")
print(head(customer_metrics[order(-customer_metrics$Total_Spent), ], 3))

Customer Analysis Summary:
Total customers analyzed: 100 
Top customer total spent: $ 8471.51 
# A tibble: 3 × 5
  CustomerID Name        Total_Spent Order_Count Avg_Order_Value
       <dbl> <chr>             <dbl>       <int>           <dbl>
1         53 Customer 53       8472.           2           4236.
2          3 Customer 3        6107.           2           3053.
3         61 Customer 61       5768.           2           2884.


In [24]:
# REQUIRED variable name: product_metrics  
# Your code here: Analyze product performance
# Required columns: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
suppressPackageStartupMessages(library(dplyr))

# Helper: rename to standard column names if variants are present
std_name <- function(df, mapping) {
  for (tgt in names(mapping)) {
    cands <- mapping[[tgt]]
    hit <- cands[cands %in% names(df)]
    if (length(hit) > 0 && !(tgt %in% names(df))) {
      names(df)[match(hit[1], names(df))] <- tgt
    }
  }
  df
}

# Normalize names (no-op if already standard)
products <- std_name(products, list(
  ProductID    = c("ProductID","Product_Id","Product_ID","product_id"),
  Product_Name = c("Product_Name","ProductName","Name","Product")
))
order_items <- std_name(order_items, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  ProductID  = c("ProductID","Product_Id","Product_ID","product_id"),
  Quantity   = c("Quantity","Qty","quantity"),
  Unit_Price = c("Unit_Price","UnitPrice","Price","UnitPriceUSD")
))
# Aggregate item lines, then left join to all products to include products with zero sales
item_summaries <- order_items %>%
  mutate(line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)) %>%
  group_by(ProductID) %>%
  summarise(
    Total_Revenue   = sum(line_total, na.rm = TRUE),
    Total_Quantity  = sum(Quantity,   na.rm = TRUE),
    Order_Frequency = n_distinct(OrderID),
    .groups = "drop"
  )

product_metrics <- products %>%
  select(ProductID, Product_Name) %>%
  left_join(item_summaries, by = "ProductID") %>%
  mutate(
    Total_Revenue   = coalesce(Total_Revenue, 0),
    Total_Quantity  = coalesce(Total_Quantity, 0),
    Order_Frequency = coalesce(Order_Frequency, 0L)
  ) %>%
  select(ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency) %>%
  arrange(desc(Total_Revenue))
# Required output for autograding:
cat("Product Analysis Summary:
")
cat("Total products analyzed:", nrow(product_metrics), "
")
cat("Top product revenue: $", round(max(product_metrics$Total_Revenue, na.rm = TRUE), 2), "
")
print(head(product_metrics[order(-product_metrics$Total_Revenue), ], 3))

Product Analysis Summary:
Total products analyzed: 50 
Top product revenue: $ 13970.66 
# A tibble: 3 × 5
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <dbl> <chr>                <dbl>          <dbl>           <int>
1        26 Product 26          13971.             60              18
2        47 Product 47          12582.             38              10
3        43 Product 43          10584.             36              10


In [25]:
# REQUIRED variable name: supplier_metrics
# Your code here: Evaluate supplier performance  
# Required columns: Supplier_ID, Supplier_Name, Product_Count

suppressPackageStartupMessages(library(dplyr))

# Helper to standardize column names if variants exist
std_name <- function(df, mapping) {
  for (tgt in names(mapping)) {
    cands <- mapping[[tgt]]
    hit <- cands[cands %in% names(df)]
    if (length(hit) > 0 && !(tgt %in% names(df))) {
      names(df)[match(hit[1], names(df))] <- tgt
    }
  }
  df
}

# Normalize expected column names (no-op if already standard)
suppliers <- std_name(suppliers, list(
  Supplier_ID   = c("Supplier_ID","SupplierId","Supplier_Id","supplier_id","SupplierID"),
  Supplier_Name = c("Supplier_Name","SupplierName","Name")
))
products <- std_name(products, list(
  ProductID   = c("ProductID","Product_Id","Product_ID","product_id"),
  Supplier_ID = c("Supplier_ID","SupplierId","Supplier_Id","supplier_id","SupplierID")
))

# Count products per supplier (include suppliers with zero products)
product_counts <- products %>%
  group_by(Supplier_ID) %>%
  summarise(Product_Count = n_distinct(ProductID), .groups = "drop")

supplier_metrics <- suppliers %>%
  select(Supplier_ID, Supplier_Name) %>%
  left_join(product_counts, by = "Supplier_ID") %>%
  mutate(Product_Count = coalesce(Product_Count, 0L)) %>%
  arrange(desc(Product_Count))

# Required output for autograding:
cat("Supplier Analysis Summary:
")
cat("Total suppliers analyzed:", nrow(supplier_metrics), "
")
cat("Max products per supplier:", max(supplier_metrics$Product_Count, na.rm = TRUE), "
")
print(supplier_metrics[order(-supplier_metrics$Product_Count), ])

Supplier Analysis Summary:
Total suppliers analyzed: 10 
Max products per supplier: 10 
# A tibble: 10 × 3
   Supplier_ID Supplier_Name Product_Count
         <dbl> <chr>                 <int>
 1           7 Supplier 7               10
 2           4 Supplier 4                6
 3          10 Supplier 10               6
 4           1 Supplier 1                5
 5           3 Supplier 3                5
 6           5 Supplier 5                5
 7           6 Supplier 6                5
 8           2 Supplier 2                3
 9           8 Supplier 8                3
10           9 Supplier 9                2


In [26]:
# REQUIRED variable name: regional_analysis
# Your code here: Create regional analysis by customer city
# Required columns: City, Total_Sales, Customer_Count, Avg_Customer_Value
suppressPackageStartupMessages(library(dplyr))

# Helper to standardize column names if variants exist
std_name <- function(df, mapping) {
  for (tgt in names(mapping)) {
    cands <- mapping[[tgt]]
    hit <- cands[cands %in% names(df)]
    if (length(hit) > 0 && !(tgt %in% names(df))) {
      names(df)[match(hit[1], names(df))] <- tgt
    }
  }
  df
}

# Normalize expected column names (no-op if already standard)
customers <- std_name(customers, list(
  CustomerID = c("CustomerID","Customer_Id","Customer_ID","customer_id","CustomerId"),
  City       = c("City","city","Customer_City","Town")
))
orders <- std_name(orders, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  CustomerID = c("CustomerID","Customer_Id","Customer_ID","customer_id","CustomerId")
))
order_items <- std_name(order_items, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  Quantity   = c("Quantity","Qty","quantity"),
  Unit_Price = c("Unit_Price","UnitPrice","Price","UnitPriceUSD")
))
regional_analysis <- customers %>%
  select(CustomerID, City) %>%
  left_join(orders %>% select(OrderID, CustomerID), by = "CustomerID") %>%
  left_join(order_items %>% select(OrderID, Quantity, Unit_Price), by = "OrderID") %>%
  mutate(line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)) %>%
  group_by(City) %>%
  summarise(
    Total_Sales     = sum(line_total, na.rm = TRUE),
    Customer_Count  = n_distinct(CustomerID),
    .groups = "drop"
  ) %>%
  mutate(Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0)) %>%
  arrange(desc(Total_Sales))
# Required output for autograding:
cat("Regional Analysis Summary:
")
cat("Total cities analyzed:", nrow(regional_analysis), "
")
cat("Highest city sales: $", round(max(regional_analysis$Total_Sales, na.rm = TRUE), 2), "
")
print(regional_analysis[order(-regional_analysis$Total_Sales), ])

Regional Analysis Summary:
Total cities analyzed: 5 
Highest city sales: $ 72277.52 
# A tibble: 5 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          72278.             24              3012.
2 Phoenix          49333.             19              2596.
3 Los Angeles      48506.             22              2205.
4 Houston          38143.             19              2008.
5 New York         36417.             16              2276.


In [27]:
# Your code here:
# Analyze product performance metrics
suppressPackageStartupMessages(library(dplyr))

stopifnot(exists("product_metrics"))

total_rev <- sum(product_metrics$Total_Revenue, na.rm = TRUE)
sold_products <- sum(coalesce(product_metrics$Total_Quantity, 0) > 0)
unsold_products <- sum(coalesce(product_metrics$Total_Quantity, 0) == 0)

product_perf <- product_metrics %>%
  mutate(Revenue_Share = ifelse(total_rev > 0, Total_Revenue / total_rev, 0)) %>%
  arrange(desc(Total_Revenue))

top10_share <- sum(head(product_perf$Revenue_Share, 10), na.rm = TRUE)

cat("Product Performance Summary:\n")
cat("Total revenue: $", round(total_rev, 2), "\n")
cat("Products with sales:", sold_products, "\n")
cat("Unsold products:", unsold_products, "\n")
cat("Top 10 products share of revenue:", round(100 * top10_share, 1), "%\n\n")

cat("Top 5 products by revenue:\n")
print(head(select(product_perf, ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency, Revenue_Share), 5))


Product Performance Summary:
Total revenue: $ 313429.5 
Products with sales: 50 
Unsold products: 0 
Top 10 products share of revenue: 44.6 %

Top 5 products by revenue:
# A tibble: 5 × 6
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <dbl> <chr>                <dbl>          <dbl>           <int>
1        26 Product 26          13971.             60              18
2        47 Product 47          12582.             38              10
3        43 Product 43          10584.             36              10
4        11 Product 11          10186.             36              12
5        32 Product 32          10105.             35              11
# ℹ 1 more variable: Revenue_Share <dbl>


In [28]:
total_customers <- nrow(customers)
total_orders    <- nrow(orders)
total_products  <- nrow(products)

pct_cust_with_orders <- 100 * (total_customers - nrow(customers_no_orders)) / total_customers
pct_orders_valid     <- 100 * (total_orders - nrow(orphaned_orders)) / total_orders
pct_products_sold    <- 100 * (total_products - nrow(products_never_ordered)) / total_products

cat(sprintf("- Join Efficiency: %.1f%% of customers placed an order; %.1f%% of orders have valid customers; %.1f%% of products were sold at least once.\n",
            pct_cust_with_orders, pct_orders_valid, pct_products_sold))

- Join Efficiency: 100.0% of customers placed an order; 80.0% of orders have valid customers; 100.0% of products were sold at least once.


In [29]:
# Your code here:
# Evaluate supplier performance
suppressPackageStartupMessages(library(dplyr))

# Helper to standardize column names
std_name <- function(df, mapping) {
  for (tgt in names(mapping)) {
    cands <- mapping[[tgt]]
    hit <- cands[cands %in% names(df)]
    if (length(hit) > 0 && !(tgt %in% names(df))) {
      names(df)[match(hit[1], names(df))] <- tgt
    }
  }
  df
}

# Normalize expected column names (no-op if already standard)
order_items <- std_name(order_items, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  ProductID  = c("ProductID","Product_Id","Product_ID","product_id"),
  Quantity   = c("Quantity","Qty","quantity"),
  Unit_Price = c("Unit_Price","UnitPrice","Price","UnitPriceUSD")
))
products <- std_name(products, list(
  ProductID   = c("ProductID","Product_Id","Product_ID","product_id"),
  Supplier_ID = c("Supplier_ID","SupplierId","Supplier_Id","supplier_id","SupplierID")
))
suppliers <- std_name(suppliers, list(
  Supplier_ID   = c("Supplier_ID","SupplierId","Supplier_Id","supplier_id","SupplierID"),
  Supplier_Name = c("Supplier_Name","SupplierName","Name")
))
# Build revenue-based supplier performance (include suppliers with zero sales)
items_with_supplier <- order_items %>%
  mutate(line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)) %>%
  inner_join(products %>% select(ProductID, Supplier_ID), by = "ProductID")

revenue_by_supplier <- items_with_supplier %>%
  group_by(Supplier_ID) %>%
  summarise(
    Total_Revenue   = sum(line_total, na.rm = TRUE),
    Products_Supplied = n_distinct(ProductID),
    Orders_Involved   = n_distinct(OrderID),
    .groups = "drop"
  )

supplier_performance <- suppliers %>%
  select(Supplier_ID, Supplier_Name) %>%
  left_join(revenue_by_supplier, by = "Supplier_ID") %>%
  mutate(
    Total_Revenue     = coalesce(Total_Revenue, 0),
    Products_Supplied = coalesce(Products_Supplied, 0L),
    Orders_Involved   = coalesce(Orders_Involved, 0L)
  ) %>%
  arrange(desc(Total_Revenue))

cat("Supplier Performance Summary:\n")
cat("Suppliers analyzed:", nrow(supplier_performance), "\n")
cat("Top supplier revenue: $", round(max(supplier_performance$Total_Revenue, na.rm = TRUE), 2), "\n")
print(head(supplier_performance, 5))

Supplier Performance Summary:
Suppliers analyzed: 10 
Top supplier revenue: $ 52328.04 
# A tibble: 5 × 5
  Supplier_ID Supplier_Name Total_Revenue Products_Supplied Orders_Involved
        <dbl> <chr>                 <dbl>             <int>           <int>
1           7 Supplier 7           52328.                10              66
2           5 Supplier 5           50692.                 5              54
3          10 Supplier 10          43450.                 6              43
4           1 Supplier 1           34821.                 5              37
5           3 Supplier 3           30401.                 5              36


In [30]:
# Your code here:
# Create regional analysis by customer city
suppressPackageStartupMessages(library(dplyr))

# Helper to standardize column names if variants exist
std_name <- function(df, mapping) {
  for (tgt in names(mapping)) {
    cands <- mapping[[tgt]]
    hit <- cands[cands %in% names(df)]
    if (length(hit) > 0 && !(tgt %in% names(df))) {
      names(df)[match(hit[1], names(df))] <- tgt
    }
  }
  df
}

# Normalize expected column names (no-op if already standard)
customers <- std_name(customers, list(
  CustomerID = c("CustomerID","Customer_Id","Customer_ID","customer_id","CustomerId"),
  City       = c("City","city","Customer_City","Town")
))
orders <- std_name(orders, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  CustomerID = c("CustomerID","Customer_Id","Customer_ID","customer_id","CustomerId")
))
order_items <- std_name(order_items, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  Quantity   = c("Quantity","Qty","quantity"),
  Unit_Price = c("Unit_Price","UnitPrice","Price","UnitPriceUSD")
))
regional_analysis <- customers %>%
  select(CustomerID, City) %>%
  left_join(orders %>% select(OrderID, CustomerID), by = "CustomerID") %>%
  left_join(order_items %>% select(OrderID, Quantity, Unit_Price), by = "OrderID") %>%
  mutate(line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)) %>%
  group_by(City) %>%
  summarise(
    Total_Sales     = sum(line_total, na.rm = TRUE),
    Customer_Count  = n_distinct(CustomerID),
    .groups = "drop"
  ) %>%
  mutate(Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0)) %>%
  arrange(desc(Total_Sales))

# Required output for autograding:
cat("Regional Analysis Summary:\n")
cat("Total cities analyzed:", nrow(regional_analysis), "\n")
cat("Highest city sales: $", round(max(regional_analysis$Total_Sales, na.rm = TRUE), 2), "\n")
print(regional_analysis[order(-regional_analysis$Total_Sales), ])

Regional Analysis Summary:
Total cities analyzed: 5 
Highest city sales: $ 72277.52 
# A tibble: 5 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          72278.             24              3012.
2 Phoenix          49333.             19              2596.
3 Los Angeles      48506.             22              2205.
4 Houston          38143.             19              2008.
5 New York         36417.             16              2276.


In [31]:
# Supplier Performance Analysis (uses complete_data created earlier)

suppressPackageStartupMessages(library(dplyr))

# Fallback: build complete_data if missing
if (!exists("complete_data") && exists("orders") && exists("order_items") && exists("products") && exists("suppliers")) {
  order_customer_items <- orders %>% left_join(order_items, by = "OrderID")
  complete_order_data  <- order_customer_items %>% left_join(products, by = "ProductID")
  complete_data        <- complete_order_data %>% left_join(suppliers, by = "Supplier_ID")
}

stopifnot(exists("complete_data"))

grp_cols <- c("Supplier_ID", "Supplier_Name")
if ("Country" %in% names(complete_data)) grp_cols <- c(grp_cols, "Country")

supplier_metrics <- complete_data %>%
  filter(!is.na(Supplier_Name)) %>%
  group_by(across(all_of(grp_cols))) %>%
  summarise(
    Total_Revenue     = sum(Quantity * Unit_Price, na.rm = TRUE),
    Products_Supplied = n_distinct(ProductID),
    Orders_Involved   = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))

cat("Supplier Performance:\n")
print(supplier_metrics)

Supplier Performance:
# A tibble: 10 × 6
   Supplier_ID Supplier_Name Country     Total_Revenue Products_Supplied
         <dbl> <chr>         <chr>               <dbl>             <int>
 1           7 Supplier 7    Germany            52328.                10
 2           5 Supplier 5    Japan              50692.                 5
 3          10 Supplier 10   Japan              43450.                 6
 4           1 Supplier 1    USA                34821.                 5
 5           3 Supplier 3    Japan              30401.                 5
 6           4 Supplier 4    South Korea        29898.                 6
 7           6 Supplier 6    Japan              27453.                 5
 8           9 Supplier 9    Japan              17144.                 2
 9           8 Supplier 8    Germany            15497.                 3
10           2 Supplier 2    USA                11746.                 3
# ℹ 1 more variable: Orders_Involved <int>


In [32]:
# Regional Analysis by Customer City
suppressPackageStartupMessages(library(dplyr))

# Ensure complete_data exists (fallback build)
if (!exists("complete_data") && all(c("orders","order_items","products","suppliers") %in% ls())) {
  order_customer_items <- orders %>% left_join(order_items, by = "OrderID")
  complete_order_data  <- order_customer_items %>% left_join(products, by = "ProductID")
  complete_data        <- complete_order_data %>% left_join(suppliers, by = "Supplier_ID")
}

stopifnot(exists("complete_data"), exists("customers"))

# Helper to standardize column names (handles .x/.y)
std_name <- function(df, mapping) {
  for (tgt in names(mapping)) {
    cands <- mapping[[tgt]]
    with_suffix <- unique(c(cands, paste0(cands, ".x"), paste0(cands, ".y")))
    hit <- with_suffix[with_suffix %in% names(df)]
    if (length(hit) > 0 && !(tgt %in% names(df))) {
      names(df)[match(hit[1], names(df))] <- tgt
    }
  }
  df
}

# Normalize needed columns
customers <- std_name(customers, list(
  CustomerID = c("CustomerID","Customer_Id","Customer_ID","customer_id","CustomerId"),
  City       = c("City","city","Customer_City","Town")
))

complete_data <- std_name(complete_data, list(
  CustomerID = c("CustomerID","Customer_Id","Customer_ID","customer_id","CustomerId"),
  Quantity   = c("Quantity","Qty","quantity"),
  Unit_Price = c("Unit_Price","UnitPrice","Price","UnitPriceUSD")
))

# Extra fallback for CustomerID regex variants
if (!("CustomerID" %in% names(complete_data))) {
  cust_cols <- grep("(?i)^customer[_ ]?id(\\.x|\\.y)?$", names(complete_data), value = TRUE)
  if (length(cust_cols) > 0) {
    names(complete_data)[match(cust_cols[1], names(complete_data))] <- "CustomerID"
  }
}

regional_analysis <- complete_data %>%
  left_join(customers %>% select(CustomerID, City), by = "CustomerID") %>%
  group_by(City) %>%
  summarise(
    Total_Sales    = sum(coalesce(Quantity, 0) * coalesce(Unit_Price, 0), na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    .groups = "drop"
  ) %>%
  mutate(Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0)) %>%
  arrange(desc(Total_Sales))

cat("Regional Analysis Summary:\n")
cat("Total cities analyzed:", nrow(regional_analysis), "\n")
cat("Highest city sales: $", round(max(regional_analysis$Total_Sales, na.rm = TRUE), 2), "\n")
print(regional_analysis)

Regional Analysis Summary:
Total cities analyzed: 6 
Highest city sales: $ 72277.52 
# A tibble: 6 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          72278.             24              3012.
2 NA               68753.              5             13751.
3 Phoenix          49333.             19              2596.
4 Los Angeles      48506.             22              2205.
5 Houston          38143.             19              2008.
6 New York         36417.             16              2276.


## Part 6: Complex Business Questions

**Tasks:**
Answer the following business questions using your joined datasets:

1. **Customer Segmentation:** Identify your top 10% of customers by total spending
2. **Product Recommendations:** Which products are frequently bought together?
3. **Supplier Dependency:** Which suppliers are most critical to the business?
4. **Market Expansion:** Which cities have high customer counts but low average order values?

Provide data-driven answers with supporting analysis.

## Part 6: Complex Business Questions

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `top_customers` - Top 10% of customers by spending (use quantile function)
2. `product_combinations` - Products frequently bought together  
3. `critical_suppliers` - Suppliers ranked by importance
4. `market_expansion` - Cities with expansion opportunities

**Expected Outputs:**
Each analysis should include specific metrics and rankings as shown in the required outputs below.

In [33]:
# REQUIRED variable name: top_customers
suppressPackageStartupMessages(library(dplyr))
stopifnot(exists("customer_metrics"),
          all(c("CustomerID","Total_Spent") %in% names(customer_metrics)))

top_10_percent_threshold <- quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE)

top_customers <- customer_metrics %>%
  filter(Total_Spent >= top_10_percent_threshold) %>%
  arrange(desc(Total_Spent))

# Required output for autograding:
cat("Customer Segmentation Results:\n")
cat("Top 10% spending threshold: $", round(top_10_percent_threshold, 2), "\n")
cat("Number of top customers:", nrow(top_customers), "\n")

Customer Segmentation Results:
Top 10% spending threshold: $ 4909.52 
Number of top customers: 10 


In [34]:
# REQUIRED variable name: product_combinations
suppressPackageStartupMessages({
  library(dplyr)
  library(tidyr)
})

stopifnot(exists("order_items"), exists("products"))

# Helper to standardize column names
std_name <- function(df, mapping) {
  for (tgt in names(mapping)) {
    cands <- mapping[[tgt]]
    hit <- cands[cands %in% names(df)]
    if (length(hit) > 0 && !(tgt %in% names(df))) {
      names(df)[match(hit[1], names(df))] <- tgt
    }
  }
  df
}

# Normalize expected column names (no-op if already standard)
order_items <- std_name(order_items, list(
  OrderID   = c("OrderID","Order_Id","Order_ID","order_id"),
  ProductID = c("ProductID","Product_Id","Product_ID","product_id")
))
products <- std_name(products, list(
  ProductID    = c("ProductID","Product_Id","Product_ID","product_id"),
  Product_Name = c("Product_Name","ProductName","Name","Product")
))

# Distinct items per order
items <- order_items %>%
  select(OrderID, ProductID) %>%
  distinct()

# Build unordered pairs within each order (no self-join)
pairs_tbl <- items %>%
  group_by(OrderID) %>%
  summarise(
    pairs = list({
      ids <- sort(unique(ProductID))
      if (length(ids) < 2) NULL else {
        m <- utils::combn(ids, 2)
        tibble(ProductID_A = m[1, ], ProductID_B = m[2, ])
      }
    }),
    .groups = "drop"
  ) %>%
  unnest(pairs)

# Count pair frequency across orders
if (is.null(pairs_tbl) || nrow(pairs_tbl) == 0) {
  product_combinations <- tibble(ProductID_A = integer(), ProductID_B = integer(), Product_A = character(), Product_B = character(), n = integer())
} else {
  product_combinations <- pairs_tbl %>%
    count(ProductID_A, ProductID_B, name = "n") %>%
    left_join(products %>% select(ProductID, Product_Name), by = c("ProductID_A" = "ProductID")) %>%
    rename(Product_A = Product_Name) %>%
    left_join(products %>% select(ProductID, Product_Name), by = c("ProductID_B" = "ProductID")) %>%
    rename(Product_B = Product_Name) %>%
    arrange(desc(n))
}

# Required output for autograding:
cat("Product Combination Analysis:\n")
cat("Total product combinations found:", nrow(product_combinations), "\n")
if (nrow(product_combinations) > 0) {
  cat("Most frequent combination appears", max(product_combinations$n, na.rm = TRUE), "times\n")
}

Product Combination Analysis:
Total product combinations found: 281 
Most frequent combination appears 3 times


In [35]:
# REQUIRED variable name: critical_suppliers
# Your code here: Analyze which suppliers are most critical
# Consider both revenue contribution and product diversity
suppressPackageStartupMessages(library(dplyr))

stopifnot(exists("order_items"), exists("products"), exists("suppliers"))

# Helper to standardize column names
std_name <- function(df, mapping) {
  for (tgt in names(mapping)) {
    cands <- mapping[[tgt]]
    hit <- cands[cands %in% names(df)]
    if (length(hit) > 0 && !(tgt %in% names(df))) {
      names(df)[match(hit[1], names(df))] <- tgt
    }
  }
  df
}

# Normalize expected column names (no-op if already standard)
order_items <- std_name(order_items, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  ProductID  = c("ProductID","Product_Id","Product_ID","product_id"),
  Quantity   = c("Quantity","Qty","quantity"),
  Unit_Price = c("Unit_Price","UnitPrice","Price","UnitPriceUSD")
))
products <- std_name(products, list(
  ProductID   = c("ProductID","Product_Id","Product_ID","product_id"),
  Supplier_ID = c("Supplier_ID","SupplierId","Supplier_Id","supplier_id","SupplierID")
))
suppliers <- std_name(suppliers, list(
  Supplier_ID   = c("Supplier_ID","SupplierId","Supplier_Id","supplier_id","SupplierID"),
  Supplier_Name = c("Supplier_Name","SupplierName","Name")
))
# Use distinct product -> supplier mapping to avoid duplication
prod_dim <- products %>% select(ProductID, Supplier_ID) %>% distinct()

# Line totals per item and attach supplier
items_with_supplier <- order_items %>%
  filter(!is.na(ProductID)) %>%
  mutate(line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)) %>%
  inner_join(prod_dim, by = "ProductID")

# Aggregate supplier metrics
revenue_by_supplier <- items_with_supplier %>%
  group_by(Supplier_ID) %>%
  summarise(
    Total_Revenue     = sum(line_total, na.rm = TRUE),
    Products_Supplied = n_distinct(ProductID),
    Orders_Involved   = n_distinct(OrderID),
    .groups = "drop"
  )

# Combine with supplier dimension; include suppliers with zero sales
critical_suppliers <- suppliers %>%
  select(Supplier_ID, Supplier_Name) %>%
  left_join(revenue_by_supplier, by = "Supplier_ID") %>%
  mutate(
    Total_Revenue     = coalesce(Total_Revenue, 0),
    Products_Supplied = coalesce(Products_Supplied, 0L),
    Orders_Involved   = coalesce(Orders_Involved, 0L)
  )
# Build importance score: 70% revenue share, 30% product diversity share
tot_rev  <- sum(critical_suppliers$Total_Revenue, na.rm = TRUE)
tot_prod <- sum(critical_suppliers$Products_Supplied, na.rm = TRUE)
critical_suppliers <- critical_suppliers %>%
  mutate(
    Revenue_Share   = ifelse(tot_rev  > 0, Total_Revenue / tot_rev, 0),
    Diversity_Share = ifelse(tot_prod > 0, Products_Supplied / tot_prod, 0),
    Importance_Score = 0.7 * Revenue_Share + 0.3 * Diversity_Share
  ) %>%
  arrange(desc(Importance_Score), desc(Total_Revenue), desc(Products_Supplied))


# Required output for autograding (uncomment after completing your analysis):
# cat("Critical Suppliers Analysis:\n")
# cat("Total suppliers analyzed:", nrow(critical_suppliers), "\n")
# print("Top 3 most critical suppliers:")
# print(head(critical_suppliers, 3))

In [36]:
# REQUIRED variable name: market_expansion  
# Your code here: Identify market expansion opportunities
# Focus on cities with multiple customers but varying order values
suppressPackageStartupMessages(library(dplyr))
stopifnot(exists("customers"), exists("orders"), exists("order_items"))

# Helper to standardize column names
std_name <- function(df, mapping) {
  for (tgt in names(mapping)) {
    cands <- mapping[[tgt]]
    hit <- cands[cands %in% names(df)]
    if (length(hit) > 0 && !(tgt %in% names(df))) {
      names(df)[match(hit[1], names(df))] <- tgt
    }
  }
  df
}

# Normalize expected columns (no-op if already standard)
customers <- std_name(customers, list(
  CustomerID = c("CustomerID","Customer_Id","Customer_ID","customer_id","CustomerId"),
  City       = c("City","city","Customer_City","Town")
))
orders <- std_name(orders, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  CustomerID = c("CustomerID","Customer_Id","Customer_ID","customer_id","CustomerId")
))
order_items <- std_name(order_items, list(
  OrderID    = c("OrderID","Order_Id","Order_ID","order_id"),
  Quantity   = c("Quantity","Qty","quantity"),
  Unit_Price = c("Unit_Price","UnitPrice","Price","UnitPriceUSD")
))

# Order totals
order_totals <- order_items %>%
  transmute(OrderID, line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)) %>%
  group_by(OrderID) %>%
  summarise(order_total = sum(line_total, na.rm = TRUE), .groups = "drop")
# Orders with city
orders_city <- orders %>%
  select(OrderID, CustomerID) %>%
  left_join(customers %>% select(CustomerID, City), by = "CustomerID")

orders_with_total <- orders_city %>%
  left_join(order_totals, by = "OrderID")

# City-level metrics
city_metrics <- orders_with_total %>%
  group_by(City) %>%
  summarise(
    Total_Sales         = sum(order_total, na.rm = TRUE),
    Order_Count         = n_distinct(OrderID),
    Customer_Count      = n_distinct(CustomerID),
    Avg_Order_Value     = ifelse(Order_Count > 0, Total_Sales / Order_Count, 0),
    Median_Order_Value  = suppressWarnings(stats::median(order_total, na.rm = TRUE)),
    Orders_per_Customer = ifelse(Customer_Count > 0, Order_Count / Customer_Count, 0),
    .groups = "drop"
  )

# Select expansion candidates: many customers, below-median AOV
min_customers <- 3L
aov_threshold <- stats::median(city_metrics$Avg_Order_Value, na.rm = TRUE)

market_expansion <- city_metrics %>%
  filter(Customer_Count >= min_customers, Avg_Order_Value <= aov_threshold) %>%
  arrange(desc(Customer_Count), Avg_Order_Value, desc(Total_Sales))


# Required output for autograding (uncomment after completing your analysis):
# cat("Market Expansion Analysis:\n")
# cat("Cities evaluated for expansion:", nrow(market_expansion), "\n")
# print("Top expansion opportunities:")
# print(head(market_expansion, 5))

In [37]:
suppressPackageStartupMessages(library(dplyr))

customer_metrics %>%
  arrange(desc(Total_Spent)) %>%
  slice(1) %>%
  transmute(`Top Customer:` = paste0(Name, " — $", round(Total_Spent, 2))) %>%
  print()
# ...existing code...

# A tibble: 1 × 1
  `Top Customer:`       
  <chr>                 
1 Customer 53 — $8471.51


In [38]:


if (!exists("customer_metrics")) {
  customer_metrics <- customers %>%
    select(CustomerID, Name) %>%
    left_join(orders %>% select(OrderID, CustomerID), by = "CustomerID") %>%
    left_join(order_items %>% select(OrderID, Quantity, Unit_Price), by = "OrderID") %>%
    mutate(line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)) %>%
    group_by(CustomerID, Name) %>%
    summarise(
      Total_Spent = sum(line_total, na.rm = TRUE),
      Order_Count = n_distinct(OrderID),
      .groups = "drop"
    ) %>%
    mutate(Avg_Order_Value = ifelse(Order_Count > 0, Total_Spent / Order_Count, 0))
}

# Build product_metrics if missing
if (!exists("product_metrics")) {
  item_summaries <- order_items %>%
    mutate(line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)) %>%
    group_by(ProductID) %>%
    summarise(
      Total_Revenue   = sum(line_total, na.rm = TRUE),
      Total_Quantity  = sum(Quantity,   na.rm = TRUE),
      Order_Frequency = n_distinct(OrderID),
      .groups = "drop"
    )

  product_metrics <- products %>%
    select(ProductID, Product_Name) %>%
    left_join(item_summaries, by = "ProductID") %>%
    mutate(
      Total_Revenue   = coalesce(Total_Revenue, 0),
      Total_Quantity  = coalesce(Total_Quantity, 0),
      Order_Frequency = coalesce(Order_Frequency, 0L)
    ) %>%
    arrange(desc(Total_Revenue))
}

# Print Top Customer and Best Product
top_cust <- customer_metrics %>% arrange(desc(Total_Spent)) %>% slice(1)
cat(sprintf("Top Customer: %s — $%.2f\n", top_cust$Name, top_cust$Total_Spent))

best_prod <- product_metrics %>% arrange(desc(Total_Revenue)) %>% slice(1)
cat(sprintf("Best Product: %s — $%.2f\n", best_prod$Product_Name, best_prod$Total_Revenue))

Top Customer: Customer 53 — $8471.51
Best Product: Product 26 — $13970.66


In [39]:
# ...existing code...
suppressPackageStartupMessages(library(dplyr))

stopifnot(exists("regional_analysis"))

if (nrow(regional_analysis) == 0) {
  cat("Regional Performance: No city data available.\n")
} else {
  top_city <- regional_analysis %>% slice_max(Total_Sales, n = 1, with_ties = FALSE)
  n3 <- min(3L, nrow(regional_analysis))
  top3 <- regional_analysis %>% slice_max(Total_Sales, n = n3, with_ties = FALSE)

  line <- sprintf(
    "Regional Performance: %s leads with $%.2f in sales across %d customers (AOV $%.2f). Top cities by sales: %s.",
    top_city$City, top_city$Total_Sales, top_city$Customer_Count, top_city$Avg_Customer_Value,
    paste(sprintf("%s ($%.2f)", top3$City, top3$Total_Sales), collapse = ", ")
  )

  cat(line, "\n")
}
# ...existing code...

Regional Performance: Chicago leads with $72277.52 in sales across 24 customers (AOV $3011.56). Top cities by sales: Chicago ($72277.52), NA ($68753.21), Phoenix ($49332.57). 


In [40]:
# ...existing code...
suppressPackageStartupMessages(library(dplyr))

stopifnot(exists("suppliers"), exists("products"))
if (!exists("order_items") || !exists("orders")) {
  stop("order_items and orders must exist")
}

# Build supplier revenue if not already available
if (exists("supplier_performance")) {
  sup_rev <- supplier_performance %>%
    select(Supplier_ID, Supplier_Name, Total_Revenue)
} else {
  items_with_supplier <- order_items %>%
    mutate(line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)) %>%
    inner_join(products %>% select(ProductID, Supplier_ID), by = "ProductID")

  sup_rev <- items_with_supplier %>%
    group_by(Supplier_ID) %>%
    summarise(Total_Revenue = sum(line_total, na.rm = TRUE), .groups = "drop") %>%
    left_join(suppliers %>% select(Supplier_ID, Supplier_Name), by = "Supplier_ID")
}

tot_rev <- sum(coalesce(sup_rev$Total_Revenue, 0), na.rm = TRUE)
n_suppliers <- n_distinct(suppliers$Supplier_ID)
n_active <- sum(coalesce(sup_rev$Total_Revenue, 0) > 0, na.rm = TRUE)

# Median products per supplier
prod_counts <- products %>% group_by(Supplier_ID) %>% summarise(Product_Count = n_distinct(ProductID), .groups = "drop")
median_products <- suppressWarnings(stats::median(prod_counts$Product_Count, na.rm = TRUE))
median_products <- ifelse(is.finite(median_products), as.integer(round(median_products)), 0L)

if (nrow(sup_rev) == 0 || tot_rev <= 0) {
  line <- sprintf(
    "Supplier Analysis: %d suppliers (%d active). No revenue recorded. Median products per supplier: %d.",
    n_suppliers, n_active, median_products
  )
} else {
  top1 <- sup_rev %>% slice_max(Total_Revenue, n = 1, with_ties = FALSE)
  top_share <- top1$Total_Revenue / tot_rev
  top3n <- min(3L, nrow(sup_rev))
  top3_share <- (sup_rev %>% slice_max(Total_Revenue, n = top3n, with_ties = FALSE) %>% summarise(s = sum(Total_Revenue, na.rm = TRUE)) %>% pull(s)) / tot_rev

  line <- sprintf(
    "Supplier Analysis: %d suppliers (%d active). Top supplier %s contributes %.1f%% of revenue; top 3 cover %.1f%%. Median products per supplier: %d.",
    n_suppliers, n_active, top1$Supplier_Name, 100 * top_share, 100 * top3_share, median_products
  )
}

cat(line, "\n")
# ...existing code...

Supplier Analysis: 10 suppliers (10 active). Top supplier Supplier 7 contributes 16.7% of revenue; top 3 cover 46.7%. Median products per supplier: 5. 


## Part 7: Summary and Insights

**Requirements for Autograding:**
Complete the analysis summary below with specific findings from your work. Your responses will be evaluated for completeness and accuracy.

**Grading Criteria:**
- Specific metrics and numbers from your analysis
- Clear identification of data quality issues  
- Actionable business recommendations
- Technical insights about join performance and appropriateness

## Analysis Summary

### Key Findings from Join Operations:
- **Join Efficiency:** [Report specific percentages of matched records]
- 100.0% of customers placed an order; 80.0% of orders have valid customers; 100.0% of products were sold at least once
  
- **Data Relationships:** [Describe the relationships you discovered between tables]
- One customer can have many orders. One order can have many items. Each item links to one product. Each product links to one supplier. Each customer belongs to one city.
  
- **Multi-table Joins:** [Explain how the data grew/changed through the join sequence]
- Joining orders to items creates one row per line item. Adding products and suppliers adds details without greatly changing row counts, except where links are missing.

### Data Quality Issues Discovered:
- **Orphaned Records:** [Report specific counts of unmatched records]
-  No customers without orders. About 20% of orders have no matching customer; all products were ordered at least once.
  
- **Missing Data:** [Identify which tables had missing or invalid references]
- Only minor missing fields. Some item quantities or prices and a few customer cities; key IDs were present.
  
- **Referential Integrity:** [Describe any foreign key violations found]
- No duplicate IDs found; main issue is orders referencing non‑existent customers.


### Business Insights:
- **Top Customer:** [Name and spending amount of highest value customer]
- Customer 53 — $8471.51
  
- **Best Product:** [Product name and revenue from your analysis]
- Product 26 — $13970.66
  
- **Regional Performance:** [Key findings about city-level sales patterns]
- Chicago leads with $72277.52 in sales across 24 customers (AOV $3011.56). Top cities by sales: Chicago ($72277.52), NA ($68753.21), Phoenix ($49332.57). 

- **Supplier Analysis:** [Insights about supplier diversity and concentration]
- 10 suppliers (10 active). Top supplier Supplier 7 contributes 16.7% of revenue; top 3 cover 46.7%. Median products per supplier: 5. 

### Strategic Recommendations:
- [Provide at least 3 specific, actionable recommendations based on your data analysis]
- Reduce orphaned orders: enforce customer ID validation at order capture and backfill missing links; target +10 pp valid orders.
-Product portfolio tuning: retire or reprice low-velocity SKUs; promote high attach-rate bundles from product_combinations to lift AOV.
-Regional experiments: run A/B tests in market_expansion cities (localized promos and bundles) to raise Avg_Order_Value and orders per customer.


### Technical Learnings:
- **Inner vs Left Joins:** [When you used each and why]
- Used inner when only matched records were needed (product pairs, supplier revenue from ordered items). Used left to retain full dimensions (customers, products, suppliers) and quantify gaps via NAs for data quality checks.
  
- **Multi-table Strategy:** [Your approach to building complex joined datasets]
- Built line level facts first (orders → order_items), then enriched with products and suppliers; for regional/customer KPIs, started from customers then joined orders and items. Selected only needed columns and used distinct keys to prevent row multiplication.
  
- **Performance Considerations:** [Observations about join efficiency and data size]
- Avoided Cartesian explosions by joining on deduplicated keys and checking row counts after each step. Precomputed line total and order totals, reused summaries, and arranged to limit memory. Used anti/semi joins for integrity diagnostics.

In [41]:
# Optional: Save your analysis results (not required for autograding)
# Your code here: Save key datasets to CSV files if desired


# Final summary output for autograding verification:
cat("\n=== HOMEWORK 6 COMPLETION SUMMARY ===\n")
cat("✓ Data Import: 5 datasets loaded\n")
cat("✓ Basic Joins: 4 join types completed\n") 
cat("✓ Multi-table Joins: 4-step progression completed\n")
cat("✓ Data Quality: Anti-joins and semi-joins performed\n")
cat("✓ Business Analysis: Customer, product, supplier, and regional analysis completed\n")
cat("✓ Complex Questions: Advanced business scenarios analyzed\n")
cat("✓ Summary: Comprehensive insights documented\n")
cat("\nAll required variables created for autograding verification.\n")


=== HOMEWORK 6 COMPLETION SUMMARY ===
✓ Data Import: 5 datasets loaded
✓ Basic Joins: 4 join types completed
✓ Multi-table Joins: 4-step progression completed
✓ Data Quality: Anti-joins and semi-joins performed
✓ Business Analysis: Customer, product, supplier, and regional analysis completed
✓ Complex Questions: Advanced business scenarios analyzed
✓ Summary: Comprehensive insights documented

All required variables created for autograding verification.
